tendermint/src/lite/verifier.rs

```diff
@@ -154,3 +165,137 @@ where
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+{
+    // can we trust the still untrusted header based on the given trusted state?
+    match check_support(
+        trusted_state,
+        untrusted_header,
+        trust_threshold,
+        trusting_period,
+        now,
+    ) {
+        Ok(_) => {
+            store.add(untrusted_header)?;
+            return Ok(());
+        }
+        Err(e) => {
+            if e != Error::InsufficientVotingPower {
+                return Err(e);
+            }
+        }
+    }
+
+    // Here we can't trust the passed in untrusted header based on the known trusted state.
+    // Run bisection: try again with a pivot header whose height is in the
+    // middle of the trusted height and the desired height (h2.height).
+
+    let trusted_header = trusted_state.last_header();
+    let trusted_height: u64 = trusted_header.header().height().value();
+    let untrusted_height: u64 = untrusted_header.header().height().value();
+    let pivot: u64 = (trusted_height + untrusted_height) / 2;
+    let pivot_header = req.signed_header(pivot)?;
+    let pivot_next_vals = req.validator_set(pivot + 1)?;
+
+    verify(&pivot_header, &pivot_next_vals)?;
+
+    // Can we trust pivot header based on trusted_state?
+    can_trust(
+        trusted_state,
+        &pivot_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    // Trust the header in between the trusted and (still) untrusted height:
+    store.add(&pivot_header)?;
+    let pivot_trusted = TS::new(pivot_header, pivot_next_vals);
+
+    // Can we trust h2 based on the (now trusted) "pivot header"?
+    can_trust(
+        &pivot_trusted,
+        untrusted_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    // Add to trusted headers:
+    store.add(untrusted_header)?;
+
+    Ok(())
+}
+
+/// This function captures the high level logic of the light client verification, i.e.,
+/// an application call to the light client module to (optionally download) and
+/// verify a header for some height.
+pub fn verify_header<TS, SH, VS, L, S, R>(
+    height: Height,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = VS>,
+    L: TrustThreshold,
+    S: Store<SignedHeader = SH>,
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+    VS: ValidatorSet,
+    SH: SignedHeader,
+{
+    // Check if we already trusted a header at the given height and it didn't expire:
+    if let Ok(sh2) = store.get(height) {
+        is_within_trust_period(sh2.header(), trusting_period, now)?
+    }
+
+    // We haven't trusted a header at height yet. Request it:
+    let sh2 = req.signed_header(height)?;
+    let sh2_next_vals = req.validator_set(height.increment())?;
+    verify(&sh2, &sh2_next_vals)?;
```

Comment:
> This actually needs to be verification against the validators for the same height, not the next height.

Whoops, will try to fix it if I find the time later!
I still think the spec should not assume the validators are included in the header when in reality there are only identifiers. Then it would be clear which validator set we are using to go from ids to validators. 

Good point about the signature verification. While we can optimize many things later, we should probably not waste sig. verification right from the start.